### DISCUSSION TITLE
Periodic tx message modify?

### DISCUSSION THREAD

**Discussion by jma1123** [cantools/cantools#274] [Q&A] [answered]
Howdy, I'm looking for suggestions and information on how to modify a periodic message using cantools.  I've created a simple test application to read in a dbc file and create a periodic message.  The python can package has a modify_data() message that I was hoping would somehow work or be supported in cantools.  I like using a database as this makes it great for defining and sharing information.  Without a modification to the message, I'm not sure the point of having a periodic message that is essentially a constant.  If nothing is supported, any recommendations?  For example use a timer, create a single instance of the message from the database and send it?  Thank you in advance.  I'm new to python and I likely overlooked something obvious in the documentation.  Cheers J -

**Comment by eerimoq** [comment]
Do you mean using the tester class? If so...

You are right, it's not possible to change the interval when the program is running. The interval is read once from the database. I guess you can modify the tester code to allow changing the interval.

Or do you mean loading a database and change the interval and dump the database to file again? Or something else?

**Comment by jma1123** [comment]
I was not clear, my apologies.  I’m looking to change the data sent while keeping the period the same.  I was not using the tester.

**Reply by andlaus** [reply]
I might misunderstand what you want to achive here here, but as far as I know, the cycle time of a message is not part of the CAN bus specification itself, i.e., it merely is an non-binding information on the bus description level that the application which sends a given message is supposed to do so with the specified frequency. Instead of using python-can's `send_periodic()` method you could send the messages yourself in your main loop (or in a separate thread/process):

```python
import can

bus = can.interface.Bus()

while True:
  time.sleep(cycle_time)
  bus.send(can.Message(arbitration_id=0x123, data=b"\x00\x01\x02"))
```

With this approach, you may change the data arbitrarily between sends. (Since I'm not really an expert w.r.t. python-can, I might be mistaken, though.)

**Comment by jma1123** [comment]
Greetings,
The encoding on the data with periodic sending is the issue.  Yes, I can use the python can library and stuff all bytes, and the use the modify function to change the data in the periodic message.  With using the cantools, I want to avoid manually creating a message very 20-200 mSec for each of the repetitive messages broadcast.  I was hoping the cantool library also had a modify function like the can library.  In my experience sending a non-changing message periodically does not accomplish much and is unfortunately not very useful.  The goal for me was to use a dbc file as it’s easy to maintain, share, and document.  The dbc file and encoding for a non-repetitive message should work fine.

**Reply by andlaus** [reply]
I'm afraid that you'll have to re-encode the entire message every time you need to change any of the data. This should come with negligible resource intensity, though. The cantools simplify this via the [`Message.encode()`](https://github.com/eerimoq/cantools/blob/master/cantools/database/can/message.py#L748) method.

**Comment by jma1123** [comment, marked as answer]
Thank you for confirming.  I’m afraid I’ll have to go a different direction .
